### ISSUE TITLE
Improved Accessibility support

### ISSUE THREAD

**Issue by knoobie** [vaadin/vaadin-messages#14]
Take this with a grain of salt, because I don't know the whole picture or design documents. 

My assumptions:
 - `vaadin-message-list` is the container for a list of `vaadin-message`
 - `vaadin-message-list` is NOT a whole chat including an input field (`vaadin-chat` incoming?)
 - `vaadin-message` can be used WITHOUT `vaadin-message-list`? 
 
 I try to sort my concerns / improvements by MUST, SHOULD; NICE TO HAVE: 
 
 MUST: 
- [x] https://github.com/vaadin/vaadin-messages/pull/15 `vaadin-message` must not have a `role=listitem` if it can be used standalone 

SHOULD:
 

- [x] https://github.com/vaadin/vaadin-messages/pull/15 `vaadin-message-list` should have a role `role=log` (https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA23)
   -  `role=log` has an implicit aria-live value of polite and aria-atomic value of false - if that doesn't work (has to be tested!) please add them to `vaadin-message-list` as well
- [x] https://github.com/vaadin/vaadin-messages/pull/15  `vaadin-message-list` should have a `aria-relevant=additions` to only read new messages 
 - `vaadin-message-list`should have a `aria-label` with a customizable string like `Chatting about Bob's cat` (i18n)
 - `vaadin-message` should help the user to identify who send a message, could be done by adding a screen reader only text like `user.name said` or `user.name said time ago:` (keep i18n in mind) 

NICE TO HAVE:
- [x] https://github.com/vaadin/vaadin-messages/pull/15  `vaadin-message-list` ~~can **not** contain the `vaadin-message` directly. Instead it should be wrapped in a `<ol>` and `<li>` so that `role=listitem` and `role=list` isn't needed anymore.~~
   - See comments in the PR, `role=listitem` and `role=list` is favoured
- [x] https://github.com/vaadin/vaadin-messages/pull/15 - `vaadin-message-list` can have a `tabindex=0` to make it focusable and highlight visually (focus-ring?) that the user focused the list of messages
-  A focused `vaadin-message-list` (which has a horizontal scrollbar) can be controlled with UP and DOWN arrow keys to read messages from the past that aren't in the view port anymore
- `vaadin-message` can follow the roving tabindex principle (https://www.w3.org/TR/wai-aria-practices/#kbd_roving_tabindex)
  - Once the user tabs **inside** `vaadin-message-list` he can rove to all `vaadin-message` with the arrow keys UP and DOWN (additional, even more nice to have: `HOME` focuses the first message, `END` the last message) 
  - Using tab again focuses the `vaadin-message-list` again or the next focusable dom element outside
- `vaadin-message` can distinguish who send the message allowing to either say `Alice said` or `You said`(keep i18n in mind)
- UX: `vaadin-message` can work with a real "time" object to allow `User said 5min ago` instead of `User said at 2PM`

**Comment by Peppe** [comment]
> - vaadin-message-list is the container for a list of vaadin-message
> - vaadin-message-list is NOT a whole chat including an input field (vaadin-chat incoming?)
> - vaadin-message can be used WITHOUT vaadin-message-list?

All the assumptions are correct. We will build a third component that we have dubbed `vaadin-message-input` The primary reason for building them, is to provide an automated chat component within Collaboration Engine (CE). It will be suitable for commenting (async), chat (sync), and activity logs. You can see the prototype of the chat component in the [CE 3.1.0 alpha2 release notes](https://github.com/vaadin/collaboration-engine/releases/tag/3.1.0.alpha2). 

Now that we have the proto, we are building the actual components as stand-alone pieces (message, message-list, message-input) in this messages repository, which can also be used separately from CE. It is unclear if we will do a component that wraps up all the three components in the messages repo, or within CE. With using the components stand alone, you would still have to take care of a lot with the various pieces, like what happens when a user clicks send in the input field and how the list of messages is populated, so we have to see how it pans out.

Thank you for all the wai aria pointers. I think we can address most of them next week, and figure out what we should do with the rest.

 If you'd like to chat (heh) about the plans, see designs, or know more, we can discuss it here in GitHub or in [Discord](https://discord.gg/MYFq5RTbBn), where we both are hanging out.

**Comment by knoobie** [comment]
> (async), chat (sync), and activity logs 

That makes the change to `role=log` event more important, good to know :) 

> prototype of the chat component in the CE 3.1.0 alpha2 release notes.

Looking at this image, my initial comment about `vaadin-message-list` should have an `aria-label` could be wrong - because the parent container could already have it. 

> It is unclear if we will do a component that wraps up all the three components in the messages repo, or within CE.  

I would prefer open source, but I can totally understand if you wanna do closed source with it. I just hope there is a wrapped component like `vaadin-chat` or `vaadin-conversation` which can wrap all components logically and semantically to leverage it for holding information like headings and so on. 

> If you'd like to chat (heh) about the plans

Well done, good pun! I think I'm good :) I'm following the development closely and I'm pretty familiar with it, because I saw the presentation in munich (Vaadin Insights 2019 (damn I'm missing events..) and we had a good discussion last year via mail about the whole CE stuff.

**Comment by anezthes** [comment]
Thanks for the feedback, knoobie!

> vaadin-message must not have a role="listitem" if it can be used standalone

For the (current set of) use cases we've been addressing so far, <vaadin-message> would/should be seen as an item in a list.

AFAIK it's not intended to be used as a standalone component. If it is, we'll have to ensure the parent component either wraps it correctly or sets the proper attribute(s).

> vaadin-message-list should have role="log"

Initially opted against role="log" because I wasn't aware of aria-relevant="additions". I'll change that. Thanks!

> vaadin-message-list should have an aria-label

I don't think anything blocks you from setting an aria-label on <vaadin-message-list>?

> vaadin-message should help the user to identify who sent a message

How beneficial is adding "said" or "at" in terms of accessibility? Most apps I've tried only seem to announce the name and time.

One thing we'll need to consider is that <vaadin-message> won't always have an author (e.g. when used in an activity log).

> vaadin-message-list can not contain the vaadin-message directly. Instead it should be wrapped in a <ol> and <li> so that role="listitem" and role="list" isn't needed anymore.

If we use role="log", <vaadin-message-list> will indeed need an internal <ol> element. However, why would you prefer generating <li> elements over setting role="listitem" on <vaadin-message>?

> Keyboard accessibility

UP and DOWN arrow keys have been discussed, but I didn't think of using HOME and END. I'll add that to the design, thank you!

IIRC, roving tab index is something we've used for other components, so it could/should be reused here as well.

> Relative time

This is on the todo list. :)

**Comment by anezthes** [comment]
Speaking of accessibility, @knoobie, I was wondering about this kind of approach that Slack, Discord, Skype etc. all have for their chats:

```
=== February 5, 2021

Patricia (19:36)
Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.

John (20:45)
Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.

=== February 6, 2021

Robert (03:07)
Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.

Jennifer (09:44):
Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
```

When a user focuses an item, the author, timestamp and message are read aloud, but not the date. Is it sufficient to just have the date as another `<li>` or should it be handled differently?

**Comment by knoobie** [comment]
@anezthes thank you for considering my feedback! 

> I don't think anything blocks you from setting an aria-label on `<vaadin-message-list>`?

You are right, nobody blocks me, but having a public API allows other people that aren't so accessibility focused to implement it easily as well - developer accessibility ;) 

 > How beneficial is adding "said" or "at" in terms of accessibility?  
 
 You can totally go with the default "User", "Time" and "Message"  but it just reads nicer e.g.:
 
`Alice 03:27 AM Lorem Ipsum` vs. `Alice said at 03:27 AM: Lorem Ipsum` 

The whole `Alice said` could be omitted if used for activity log without a user.  It's more of a UX thing for SR user. 

>  Any particular reason why you prefer `<li>` elements over setting role=listitem"?

**you should only use WAI-ARIA when you need to!**  

Source: https://developer.mozilla.org/en-US/docs/Learn/Accessibility/WAI-ARIA_basics#when_should_you_use_wai-aria

If possible, always use the native HTML elements because they are way better supported by AT instead of "recreating" them. 

> but I didn't think of using HOME and END

You can find those hot keys in most of the list, grids, menus and other components listed here https://www.w3.org/TR/wai-aria-practices - it's often optional but allows for fast navigation. 

> Is it sufficient to just have the date as another `<li>` or should it be handled differently?

In the end, this "message" is just a special styled "log activity" - so creating a separate `<li>` sounds totally reasonable for me. Things to consider: the roving tabindex should be done on the `<li>` instead of the `<vaadin-message>` to announce this messages while roving as well that aren't wrapped by a `<vaadin-message>` if you don't wanna put those in a separate `<vaadin-message>`.

**Comment by Peppe** [comment]
#22 Fixes many of the standing topics in this issue.
